client/cache.go

```diff
@@ -0,0 +1,63 @@
+package client
+
+import (
+	"github.com/hashicorp/vault/api"
+	"sync"
+)
+
+// Cache is a thread-safe cache for vault queries
+type Cache struct {
+	vaultClient *api.Client
+	listQueries map[string]*cacheElement
+	mutex       sync.Mutex
+}
+
+type cacheElement struct {
+	Secret *api.Secret
+	Err    error
+}
+
+// NewCache create a new thread-safe cache object
+func NewCache(vaultClient *api.Client) *Cache {
+	return &Cache{
+		vaultClient: vaultClient,
+		listQueries: make(map[string]*cacheElement),
+	}
+}
+
+// Clear removes all entries from the cache
+func (cache *Cache) Clear() {
+	cache.mutex.Lock()
+	cache.listQueries = make(map[string]*cacheElement)
+	cache.mutex.Unlock()
+}
+
+// List tries to get path from cache.
+// If path is not available in cache, it uses the vault client to query and update cache.
+func (cache *Cache) List(path string) (result *api.Secret, err error) {
+	// try to get path from cache
+	cache.mutex.Lock()
+	hit := false
+	if value, ok := cache.listQueries[path]; ok {
+		result = value.Secret
+		err = value.Err
+		hit = true
```

Comment:
Can't `ok` be used for this?

Reply:
Good catch :+1: 
Adjusted it now

Reply:
Sorry, apparently this doesn't work:

```
var hit bool
if value, hit := cache.listQueries[path]; hit {
    result = value.Secret
    err = value.Err
}
if hit {
    // this branch is never reached
    return result, err
}
```
Just now verified that it results in `hit` always being false after leaving the first if scope. I am not really sure why though

Reply:
Maybe the assignment is considered in the scope of the `if` because of the `:=`? These little Golang things catch me out too. :)

Reply:
Now this works :)

```
value, hit := cache.listQueries[path]
if hit {
    return value.Secret, value.Err
}
```

I think this part `; ok {` binds the two initialized variables solely to the if scope.